### PR TITLE
Fix inconsistent outputs in `on_*_end` and `*_end`

### LINKED ISSUE #6921: Inconsistent `outputs` format between `training_epoch_end` and `on_train_epoch_end`
## 🐛 Bug

The outputs object for `on_train_epoch_end` should not include the `extra` field

### To Reproduce

```python
def test_bug(tmpdir):
    class TestModel(BoringModel):
        def training_step(self, batch, batch_idx):
            output = self(batch)
            loss = self.loss(batch, output)
            return {"loss": loss, "foo": 123}

        def training_epoch_end(self, outputs):
            print("training_epoch_end:", outputs)

        def on_train_epoch_end(self, outputs):
            print("on_train_epoch_end:", outputs)

    class TestCallback(Callback):
        def on_train_epoch_end(self, trainer, pl_module, outputs):
            print("callback on_train_epoch_end:", outputs)

    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=True, callbacks=[TestCallback()], progress_bar_refresh_rate=0)
    trainer.fit(TestModel())
```

```console
callback on_train_epoch_end: [[[{'extra': {'foo': 123}, 'minimize': tensor(1.1792)}]]]
on_train_epoch_end: [[[{'extra': {'foo': 123}, 'minimize': tensor(1.1792)}]]]
training_epoch_end: [{'foo': 123, 'loss': tensor(1.1792)}]
```

### Expected behavior

`on_train_epoch_end` acts as `training_epoch_end`

### Environment

master

### Additional context

Reported by Marek O in Slack

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Fixes #6921

Several changes:
- Move `training_epoch_end` (which actual handles model epoch end) from logger connector to `training_loop.py`
- Use one static method (`_prepare_outputs`) in `training_loop.py` to unpack all `Result` object outputs from both `epoch_end` and `batch_end`
- Pass these unpacked outputs to __both__ `on_train_epoch_end` and `training_epoch_end` hooks (previously, `training_epoch_end` behaved differently)
- Moved evaluation output handling to `trainer.py` so that both `evaluation_epoch_end` and `on_evaluation_epoch_end` get the same `outputs` argument
- Changed order of calls to `training_epoch_end` and `on_train_epoch_end` to be consistent with the docs

Previously, running the code from the issue, the user saw this output:

```python
callback on_validation_batch_end: {'loss': tensor(1.9916), 'foo': 'from_val_step'}
validation_epoch_end: [{'loss': tensor(1.9916), 'foo': 'from_val_step'}]
callback on_validation_epoch_end: [[{'loss': tensor(1.9916), 'foo': 'from_val_step'}]]
on_validation_epoch_end: [[{'loss': tensor(1.9916), 'foo': 'from_val_step'}]]
callback on_train_batch_end: [[{'extra': {'foo': 123}, 'minimize': tensor(0.2295)}]]
callback on_train_epoch_end: [[[{'extra': {'foo': 123}, 'minimize': tensor(0.2295)}]]]
on_train_epoch_end: [[[{'extra': {'foo': 123}, 'minimize': tensor(0.2295)}]]]
training_epoch_end: [{'foo': 123, 'loss': tensor(0.2295)}]
callback on_validation_batch_end: {'loss': tensor(2.2656), 'foo': 'from_val_step'}
validation_epoch_end: [{'loss': tensor(2.2656), 'foo': 'from_val_step'}]
callback on_validation_epoch_end: [[{'loss': tensor(2.2656), 'foo': 'from_val_step'}]]
on_validation_epoch_end: [[{'loss': tensor(2.2656), 'foo': 'from_val_step'}]]
```

Now sees this:
```python
callback on_validation_batch_end: {'loss': tensor(0.6422), 'foo': 'from_val_step'}
validation_epoch_end: [{'loss': tensor(0.6422), 'foo': 'from_val_step'}]
callback on_validation_epoch_end: [{'loss': tensor(0.6422), 'foo': 'from_val_step'}]
on_validation_epoch_end: [{'loss': tensor(0.6422), 'foo': 'from_val_step'}]
callback on_train_batch_end: {'foo': 123, 'loss': tensor(1.2562)}
training_epoch_end: [{'foo': 123, 'loss': tensor(1.2562)}]
callback on_train_epoch_end: [{'foo': 123, 'loss': tensor(1.2562)}]
on_train_epoch_end: [{'foo': 123, 'loss': tensor(1.2562)}]
callback on_validation_batch_end: {'loss': tensor(0.4613), 'foo': 'from_val_step'}
validation_epoch_end: [{'loss': tensor(0.4613), 'foo': 'from_val_step'}]
callback on_validation_epoch_end: [{'loss': tensor(0.4613), 'foo': 'from_val_step'}]
on_validation_epoch_end: [{'loss': tensor(0.4613), 'foo': 'from_val_step'}]
```

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by ananthsub** [comment]
@ethanwharris n00b question: should the lightning module hook come before the callback hook runs?

**Comment by ethanwharris** [comment]
@ananthsub The docs certainly think it should, so it will following this PR :smiley:

File: CHANGELOG.md [hunk 1]
```diff
@@ -240,6 +240,15 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed `AttributeError for `require_backward_grad_sync` when running manual optimization with sharded plugin ([#6915](https://github.com/PyTorchLightning/pytorch-lightning/pull/6915))
 
 
+- Fixed a bug where `LightningModule.train_epoch_end` was called after the `on_train_end_epoch` hook ([#6969](https://github.com/PyTorchLightning/pytorch-lightning/pull/6969))
```
Comment (ananthsub) [inline CHANGELOG.md:243]:
```suggestion
- Fixed a bug where `LightningModule.training_epoch_end` was called after the `on_train_end_epoch` hook ([#6969](https://github.com/PyTorchLightning/pytorch-lightning/pull/6969))
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -240,6 +240,15 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed `AttributeError for `require_backward_grad_sync` when running manual optimization with sharded plugin ([#6915](https://github.com/PyTorchLightning/pytorch-lightning/pull/6915))
 
 
+- Fixed a bug where `LightningModule.training_epoch_end` was called after the `on_train_end_epoch` hook ([#6969](https://github.com/PyTorchLightning/pytorch-lightning/pull/6969))
+
+
+- Fixed a bug where the outputs object passed to `LightningModule.train_epoch_end` was different from the object passed to the `on_train_end_epoch` hook ([#6969](https://github.com/PyTorchLightning/pytorch-lightning/pull/6969))
+
+
+- Fixed a bug where the outputs passed to `train_batch_end` would be lists even when using a single optimiser and no truncated backprop through time steps ([#6969](https://github.com/PyTorchLightning/pytorch-lightning/pull/6969))
```
Comment (carmocca) [inline CHANGELOG.md:249]: Triple kill! 💥

**Review by carmocca — COMMENTED** [review]
Maybe add a test for the expected outputs format?

**Comment by carmocca** [comment]
Another idea would be to merge together these two tests:

https://github.com/PyTorchLightning/pytorch-lightning/blob/fe0d08899eba94d275ff42253f495d9e70d86f89/tests/models/test_hooks.py#L261

https://github.com/PyTorchLightning/pytorch-lightning/blob/fe0d08899eba94d275ff42253f495d9e70d86f89/tests/callbacks/test_callbacks.py#L22

So we can properly assert the order for LightningModule's hooks (base ones and `on_*`) + Callback's hooks (`on_*`)

File: pytorch_lightning/trainer/evaluation_loop.py [hunk 3]
```diff
@@ -205,46 +205,40 @@ def log_epoch_metrics_on_evaluation_end(self):
...
         # capture logging
         self.trainer.logger_connector.cache_logged_metrics()
         # depre warning
-        if eval_results is not None and user_reduced:
+        if epoch_end_output is not None and user_reduced:
             step = 'testing_epoch_end' if self.trainer.testing else 'validation_epoch_end'
             self.warning_cache.warn(
                 f'The {step} should not return anything as of 9.1.'
                 ' To log, use self.log(...) or self.write(...) directly in the LightningModule'
             )
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/evaluation_loop.py:235]: Probably a typo for `0.9.1`

Deprecate what exactly?

[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/trainer/evaluation_loop.py:235]: storing the value of `model.validation_epoch_end` or `model.test_epoch_end` and checking if it's not None

[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/evaluation_loop.py:235]: Ah yes. This is so old it should be fine to just remove it considering we had the warning.

Feel free to do it in a separate PR

[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/trainer/evaluation_loop.py:235]: https://github.com/PyTorchLightning/pytorch-lightning/pull/6973

File: CHANGELOG.md [hunk 4]
```diff
@@ -240,6 +240,15 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed `AttributeError for `require_backward_grad_sync` when running manual optimization with sharded plugin ([#6915](https://github.com/PyTorchLightning/pytorch-lightning/pull/6915))
 
 
+- Fixed a bug where `LightningModule.training_epoch_end` was called after the `on_train_end_epoch` hook ([#6969](https://github.com/PyTorchLightning/pytorch-lightning/pull/6969))
```
Comment (Borda) [inline CHANGELOG.md:243]: can we write these changes more densely, this is quite long and someone needs detail he can open the particular PR, see the chlog entry as kind of title/subtitle

**Review by Borda — COMMENTED** [review]
looks good, pls add typing

File: CHANGELOG.md (hunk 4 above)
Comment (carmocca) [reply CHANGELOG.md:243]: I personally think it's nicer to mention what was exactly fixed than just saying "Fix inconsistent outputs in on_*_end and *_end"

File: pytorch_lightning/trainer/training_loop.py [hunk 5]
```diff
@@ -504,15 +556,48 @@ def run_training_epoch(self):
         # progress global step according to grads progress
         self.increment_accumulated_grad_global_step()
 
+    def on_train_epoch_end(self, epoch_output):
```
Comment (ananthsub) [inline pytorch_lightning/trainer/training_loop.py:559]:
```suggestion
    def on_train_epoch_end(self, epoch_output) -> None:
```

**Review by ananthsub — APPROVED** [review]

File: CHANGELOG.md [hunk 6]
```diff
@@ -240,6 +240,15 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
+
+
+- Fixed a bug where the outputs object passed to `LightningModule.training_epoch_end` was different from the object passed to the `on_train_end_epoch` hook ([#6969](https://github.com/PyTorchLightning/pytorch-lightning/pull/6969))
+
+
+- Fixed a bug where the outputs passed to `train_batch_end` would be lists even when using a single optimiser and no truncated backprop through time steps ([#6969](https://github.com/PyTorchLightning/pytorch-lightning/pull/6969))
```
Comment (awaelchli) [inline CHANGELOG.md:249]:
```suggestion
- Fixed a bug where the outputs passed to `train_batch_end` would be lists even when using a single optimizer and no truncated backprop through time steps ([#6969](https://github.com/PyTorchLightning/pytorch-lightning/pull/6969))
```

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/training_loop.py [hunk 7]
```diff
@@ -342,6 +346,58 @@ def _process_training_step_output(self, training_step_output, split_batch):
...
+
+            for batch_outputs in opt_outputs:
+                processed_tbptt_outputs = []
+
+                for tbptt_output in batch_outputs:
+                    out = tbptt_output.extra
+                    out['loss'] = tbptt_output.minimize
+                    processed_tbptt_outputs.append(out)
+
+                # if there was only one tbptt step then we can collapse that dimension
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/training_loop.py:386]: Appreciate the comments!

**Review by kaushikb11 — APPROVED** [review]

**Comment by ethanwharris** [comment]
@carmocca Have added some tests :smiley: - one for call order, and one checking that they are called with the correct outputs objects.

File: pytorch_lightning/trainer/training_loop.py [hunk 8]
```diff
@@ -342,6 +346,58 @@ def _process_training_step_output(self, training_step_output, split_batch):
...
+            batch_mode: bool,
+    ) -> Union[List[List[List[Dict]]], List[List[Dict]], List[Dict], Dict]:
+        """
+        Extract required information from batch or epoch end results.
+
+        Args:
+            outputs: A 3-dimensional list of ``Result`` objects with dimensions:
+                [optimizer outs][batch outs][tbptt steps].
+
+            batch_mode: If True, ignore the batch output dimension
```
Comment (ethanwharris) [inline pytorch_lightning/trainer/training_loop.py:361]:
```suggestion
            batch_mode: If True, ignore the batch output dimension.
```

File: tests/trainer/test_evaluation_loop.py [hunk 9]
```diff
@@ -17,7 +17,7 @@
 from tests.helpers.boring_model import BoringModel
 
 
-@mock.patch("pytorch_lightning.trainer.evaluation_loop.EvaluationLoop.call_on_evaluation_epoch_end_hook")
+@mock.patch("pytorch_lightning.trainer.evaluation_loop.EvaluationLoop.on_evaluation_epoch_end")
 def test_call_on_evaluation_epoch_end_hook(eval_epoch_end_mock, tmpdir):
     """
     Tests that `call_on_evaluation_epoch_end_hook` is called
```
Comment (carmocca) [inline tests/trainer/test_evaluation_loop.py:23]: Update the docstring and function name?

**Review by carmocca — APPROVED** [review]

File: tests/trainer/test_evaluation_loop.py (hunk 9 above)
Comment (ethanwharris) [reply tests/trainer/test_evaluation_loop.py:23]: Haha, whoops

Comment (ethanwharris) [reply tests/trainer/test_evaluation_loop.py]: Done :smiley:

**Review by tchaton — APPROVED** [review]
LGTM !
